processor/batchprocessor/batch_processor.go

```diff
@@ -0,0 +1,223 @@
+func (bp *batchProcessor) Start(ctx context.Context, host component.Host) error {
+	return nil
+}
+
+// Shutdown is invoked during service shutdown.
+func (bp *batchProcessor) Shutdown(context.Context) error {
+	// TODO: flush accumulated data.
+	return nil
+}
+
+func (bp *batchProcessor) sendItems(
+	itemsToProcess []data.TraceData,
+	resourceSpansCount uint32,
+	measure *stats.Int64Measure,
+) {
+	td := data.NewTraceData()
+	rss := td.ResourceSpans()
+
+	// TODO: Use resourceSpansCount to set capacity of the rss slice
+	// For now there is no way to do that with internal data structure API
+
+	for _, td := range itemsToProcess {
+		td.ResourceSpans().MoveTo(rss)
+	}
+
+	statsTags := []tag.Mutator{tag.Insert(processor.TagProcessorNameKey, bp.name)}
+	_ = stats.RecordWithTags(context.Background(), statsTags, measure.M(1))
+
+	// TODO: This process should be done in an async way, perhaps with a channel + goroutine worker(s)
+	_ = bp.sender.ConsumeTrace(context.Background(), td)
+}
+
+type batch struct {
+	parent                  *batchProcessor
+	mu                      sync.RWMutex
+	items                   []data.TraceData
+	totalResourceSpansCount uint32
+	totalSpanCount          uint32
+	lastSent                int64
+}
+
+func newBatch(parent *batchProcessor) *batch {
+	return &batch{
+		parent: parent,
+		items:  make([]data.TraceData, 0, initialBatchCapacity),
+	}
+}
+
+// add updates current batch by adding new TraceData object
+func (b *batch) add(td data.TraceData) {
+	newResourceSpansCount := td.ResourceSpans().Len()
+	if newResourceSpansCount == 0 {
+		return
+	}
+
+	var itemsToProcess []data.TraceData
+	var totalResourceSpansCount uint32
+
+	b.mu.Lock()
+	b.items = append(b.items, td)
+	totalSpanCount := b.totalSpanCount + uint32(td.SpanCount())
+	b.totalSpanCount = totalSpanCount
+	b.totalResourceSpansCount = b.totalResourceSpansCount + uint32(newResourceSpansCount)
+
+	if totalSpanCount >= b.parent.sendBatchSize {
+		itemsToProcess, totalResourceSpansCount = b.getAndReset()
+	}
+	b.mu.Unlock()
+
+	if totalResourceSpansCount > 0 {
+		b.parent.sendItems(itemsToProcess, totalResourceSpansCount, statBatchSizeTriggerSend)
+	}
+}
+
+// getAndReset resent the current batch and returns its items that should be processed
+func (b *batch) getAndReset() ([]data.TraceData, uint32) {
+	itemsToProcess := b.items
+	totalResourceSpansCount := b.totalResourceSpansCount
+	b.items = make([]data.TraceData, 0, len(itemsToProcess))
+	b.lastSent = time.Now().UnixNano()
+	b.totalSpanCount = 0
+	b.totalResourceSpansCount = 0
+	return itemsToProcess, totalResourceSpansCount
+}
+
+type ticker struct {
+	ticker *time.Ticker
+	parent *batchProcessor
+	stopCn chan struct{}
+	once   sync.Once
+}
+
+func newTicker(parent *batchProcessor) *ticker {
+	return &ticker{
+		ticker: time.NewTicker(parent.tickTime),
+		parent: parent,
+		stopCn: make(chan struct{}),
+	}
+}
+
+func (bt *ticker) start() {
+	bt.once.Do(bt.runTicker)
+}
+
+func (bt *ticker) runTicker() {
```

Comment:
I am not sure I like this implementation.

The ticker runs independently from batch sending operation. We periodically check if time is out and we need to send the batch.

Why don't we simply start a timer after we send a batch and reset the timer after every send? I think it would result in simpler code and better behavior. They way it is implemented now will result in less unpredictable time after lastSent.

It would also allow to get rid of `tick_time` config parameter.

Reply:
I like the suggested approach. Thanks. I'll change it

Reply:
done